<a href="https://colab.research.google.com/github/integer-class/machine-learning/blob/main/AL_AZHAR_RRF/jobsheet_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab Work 1
## Data Imputation
In this work, you will attempt to perform data imputation on missing values. The dataset used for this purpose is the Titanic dataset.

### Step 1 - Load Data

Import pandas and load data to pandas's DataFrame.

In [7]:
import pandas as pd

data = 'Titanic-Dataset.csv' # path dataset
df = pd.read_csv(data) # load dataset

Show the 5 first data by using df.head()

In [23]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  DECK        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  DECK        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  DECK        S

### Step 2 - Data Checking

In this step, you will inspect the data before performing data imputation. You can get general information about the data by using df.info() command.

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


From df.info() we get the information about the number of non null data for each column. Dispite of that, we also know that,

Age type is float64

Fare type is float64

Embarked type is object.

In Pandas, object can be a string object. You need to ensure the type of object!

Next, we will check the total null data from these columns by using df.isnull().sum() command.

In [10]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

From that information, we know that,

The total of null values of Age is 177.

The total of null values of Cabin is 687.

The total of null values of Embarked is 2.

### Step 3 - Imputation

In this step, we will perform imputation for missing data in "Age," "Cabin," and "Embarked." The strategy we will use is as follows,

"Age" --> Since "Age" is nominal data, we will use the mean strategy.

"Cabin" --> "Cabin" is related to passenger cabin numbers. Here, we will assume that all passengers without cabin numbers stayed on the ship's decks. Therefore, we will substitute missing data with the information "DECK."

"Embarked" --> "Embarked" represents the location of passenger embarkation. There are only 2 missing data points here. Since the data is nominal, we can use the mode to substitute missing data.

In [12]:
# Age - mean
df['Age'].fillna(value=df['Age'].mean(), inplace=True)

# Cabin - "DECK"
df['Cabin'].fillna(value="DECK", inplace=True)

# Embarked - modus
df['Embarked'].fillna(value=df['Embarked'].mode, inplace=True)

### Step 4 - Verification

Perform a recheck with df.info(). Furthermore, we can also verify whether the filled values are correct by directly checking the data. The results of the imputation process are as follows,

In [18]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          891 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        891 non-null    object 
 11  Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [25]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  DECK        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  DECK        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  DECK        S

Full Code
"Age" --> Since "Age" is nominal data, we will use the mean strategy. "Cabin" --> "Cabin" is related to passenger cabin numbers. Here, we will assume that all passengers without cabin numbers stayed on the ship's decks. Therefore, we will substitute missing data with the information "DECK." "Embarked" --> "Embarked" represents the location of passenger embarkation. There are only 2 missing data points here. Since the data is nominal, we can use the mode to substitute missing data.

# Lab Work 2
Feature Extraction, Encoding, Standardization

### Feature Extraction
In this stage, we will select which variables we will use as features. In the foundational theory section, we have already learned that not all variables can always be used as features. Therefore, we can choose features that are deemed important for the purpose of creating a machine learning model. We will use the same case, which is to identify the characteristics of Titanic passengers who survived. Thus, the features we will use are,

"Survived"

"Pclass"

"Sex"

"Age"

"Cabin"

We will separate these variables from the others. The data we will use is the Titanic data that has been repaired through the imputation process.

### Step 0 - Load Library

In [26]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler

### Step 1 - Load Data

In [28]:
dpath = 'Titanic-Dataset-fixed.csv'
df = pd.read_csv(dpath)
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  DECK        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  DECK        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500  DECK        S

### Step 2 - Slice The Data

We will select only the columns that we think is features.

In [29]:
df = df[['Survived', 'Pclass', 'Age', 'Sex', 'Cabin']]
df.head()

Survived  Pclass   Age     Sex Cabin
0         0       3  22.0    male  DECK
1         1       1  38.0  female   C85
2         1       3  26.0  female  DECK
3         1       1  35.0  female  C123
4         0       3  35.0    male  DECK

### Step 3 - Encoding

We will perform encoding on categorical data, namely "Sex" and "Cabin." This step is necessary because machine learning algorithms perform mathematical calculations. The strategy we will use is Label Encoding. Label Encoding is similar to Ordinal Encoding, with the only difference being the sorting process. Label Encoding does not sort the data beforehand.

In [30]:
le = LabelEncoder() # initiate LabelEncoder object
df['Sex'] = le.fit_transform(df['Sex']) # encoding
df['Cabin'] = le.fit_transform(df['Cabin']) # encoding

### Step 4 - Verification

Rerun df.head() command to verify the result.

In [31]:
df.head()

Survived  Pclass   Age  Sex  Cabin
0         0       3  22.0    1    115
1         1       1  38.0    0     81
2         1       3  26.0    0    115
3         1       1  35.0    0     55
4         0       3  35.0    1    115

### Step 5 - Standardization

In this stage, we will standardize "Age." This is not actually necessary because the values in "Age" are not too far apart. The purpose of this step is to demonstrate the standardization process.

In [32]:
std = StandardScaler()
df['Age'] = std.fit_transform(df[['Age']])

### Step 6 - Verification

Rerun df.head() command to verify the result.

In [33]:
df.head()

Survived  Pclass       Age  Sex  Cabin
0         0       3 -0.592481    1    115
1         1       1  0.638789    0     81
2         1       3 -0.284663    0    115
3         1       1  0.407926    0     55
4         0       3  0.407926    1    115

# Lab Work 3

Splitting Data

In this lab work, we will try three different strategies for creating training, validation, and test data. The data used will be the Titanic ship data for which we extracted features in the previous practical session.

## Random Split

### Step 1 - Load Data

In [35]:
import pandas as pd

df = pd.read_csv('Titanic-Dataset-selected.csv')
df.head()

Survived  Pclass       Age  Sex  Cabin
0         0       3 -0.592481    1    115
1         1       1  0.638789    0     81
2         1       3 -0.284663    0    115
3         1       1  0.407926    0     55
4         0       3  0.407926    1    115

### Step 2 - Split Data

In [36]:
# Split data
from sklearn.model_selection import train_test_split

# Split data training dan dan lainnya
# data lainnya, akan kita split lagi menjadi validasi dan testing.
# Rasio yang akan kita gunakan adalah 8:1:1
df_train, df_unseen = train_test_split(df, test_size=0.2, random_state=0)

# Split lagi antara validasi dan testing
df_val, df_test = train_test_split(df_unseen, test_size=0.5, random_state=0)

# Cek masing-masing ukuran data

print(f'Jumlah data asli: {df.shape[0]}')
print(f'Jumlah data train: {df_train.shape[0]}')
print(f'Jumlah data val: {df_val.shape[0]}')
print(f'Jumlah data test: {df_test.shape[0]}')

# Cek rasio tiap label
print('=========')
print(f'Jumlah label data asli:\n{df.Survived.value_counts()}')
print(f'Jumlah label data train:\n{df_train.Survived.value_counts()}')
print(f'Jumlah label data val:\n{df_val.Survived.value_counts()}')
print(f'Jumlah label data test:\n{df_test.Survived.value_counts()}')

Jumlah data asli: 891
Jumlah data train: 712
Jumlah data val: 89
Jumlah data test: 90
Jumlah label data asli:
Survived
0    549
1    342
Name: count, dtype: int64
Jumlah label data train:
Survived
0    439
1    273
Name: count, dtype: int64
Jumlah label data val:
Survived
0    53
1    36
Name: count, dtype: int64
Jumlah label data test:
Survived
0    57
1    33
Name: count, dtype: int64


## Stratified Split

### Step 1 - Load Data

In [38]:
import pandas as pd

df2 = pd.read_csv('Titanic-Dataset-selected.csv')
df2.head()

Survived  Pclass       Age  Sex  Cabin
0         0       3 -0.592481    1    115
1         1       1  0.638789    0     81
2         1       3 -0.284663    0    115
3         1       1  0.407926    0     55
4         0       3  0.407926    1    115

### Step 2 - Split Data

In [39]:
# Split data
from sklearn.model_selection import train_test_split

# Split data training dan dan lainnya
# data lainnya, akan kita split lagi menjadi validasi dan testing.
# Rasio yang akan kita gunakan adalah 8:1:1
df2_train, df2_unseen = train_test_split(df2, test_size=0.2, random_state=0, stratify=df['Survived'])

# Split lagi antara validasi dan testing
df2_val, df2_test = train_test_split(df2_unseen, test_size=0.5, random_state=0, stratify=df_unseen['Survived'])

# Cek masing-masing ukuran data

print(f'Jumlah label data asli:\n{df2.Survived.value_counts()}')
print(f'Jumlah label data train:\n{df2_train.Survived.value_counts()}')
print(f'Jumlah label data val:\n{df2_val.Survived.value_counts()}')
print(f'Jumlah label data test:\n{df2_test.Survived.value_counts()}')

Jumlah label data asli:
Survived
0    549
1    342
Name: count, dtype: int64
Jumlah label data train:
Survived
0    439
1    273
Name: count, dtype: int64
Jumlah label data val:
Survived
0    51
1    38
Name: count, dtype: int64
Jumlah label data test:
Survived
0    59
1    31
Name: count, dtype: int64


## Cross Validation 1

In this work, we will employ cross validation method to split the data. Futhermore, we only create training and test data.

### Step 1 - Load Data

In [41]:
import pandas as pd

df3 = pd.read_csv('Titanic-Dataset-selected.csv')
df3.head()

Survived  Pclass       Age  Sex  Cabin
0         0       3 -0.592481    1    115
1         1       1  0.638789    0     81
2         1       3 -0.284663    0    115
3         1       1  0.407926    0     55
4         0       3  0.407926    1    115

### Step 2 - Split Data

In [42]:
# Implementasi k-fold cross validation (random) dengan training dan testing saja
from sklearn.model_selection import KFold

# inisiasi obyek kfold
kf = KFold(n_splits=4)
print(f'Jumlah fold: {kf.get_n_splits()}')
print(f'Obyek KFold: {kf}')

# Lakukan splitting dengan KFold
kf_split = kf.split(df3)
print(f'Jumlah data df: {df.shape[0]}')

# cek index data tiap fold
for train_index, test_index in kf_split:
    print(f'Index train: {train_index} | Index test: {test_index}')

Jumlah fold: 4
Obyek KFold: KFold(n_splits=4, random_state=None, shuffle=False)
Jumlah data df: 891
Index train: [223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240
 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258
 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276
 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294
 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312
 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330
 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348
 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366
 367 368 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384
 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401 402
 403 404 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420
 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438
 439 440 44

Pay attention to the index numbers for each fold. If successful, each fold should have different index numbers for both training and test data.

## Cross Validation 2

In this work, we are taking a slightly different approach compared to the previous work. First, we will split the data into training and test data. Then, we will further split the training data into several parts using cross-validation for training and validation data purposes.

### Step 1 - Load Data

In [43]:
import pandas as pd

df4 = pd.read_csv('Titanic-Dataset-selected.csv')
df4.head()

Survived  Pclass       Age  Sex  Cabin
0         0       3 -0.592481    1    115
1         1       1  0.638789    0     81
2         1       3 -0.284663    0    115
3         1       1  0.407926    0     55
4         0       3  0.407926    1    115

### Step 1 - Split Data

In [44]:
# Implementasi k-fold cross validation (random) dengan training, validation, dan testing data
from sklearn.model_selection import train_test_split, KFold

# Split dulu antara data training dan testing dengan train_test_split
# Rasio 8:2 untuk training dan testing
df4_train, df4_test = train_test_split(df4, test_size=0.2, random_state=0)

# inisiasi obyek kfold
kf2 = KFold(n_splits=4)
print(f'Jumlah fold: {kf2.get_n_splits()}')
print(f'Obyek KFold: {kf2}')

# Lakukan splitting dengan KFold untuk data df_training
# Dengan acara ini, kita masih memiliki data testing untuk keperluan pengujian model
# namun tetap dapat melakukan evaluasi dengan menggunakan data validasi
kf2_split = kf2.split(df_train)
print(f'Jumlah data df_train: {df4_train.shape[0]}')

# cek index data tiap fold
for train_index, test_index in kf2_split:
    print(f'Index train: {train_index} | Index test: {test_index}')

Jumlah fold: 4
Obyek KFold: KFold(n_splits=4, random_state=None, shuffle=False)
Jumlah data df_train: 712
Index train: [178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195
 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213
 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231
 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249
 250 251 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267
 268 269 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285
 286 287 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303
 304 305 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321
 322 323 324 325 326 327 328 329 330 331 332 333 334 335 336 337 338 339
 340 341 342 343 344 345 346 347 348 349 350 351 352 353 354 355 356 357
 358 359 360 361 362 363 364 365 366 367 368 369 370 371 372 373 374 375
 376 377 378 379 380 381 382 383 384 385 386 387 388 389 390 391 392 393
 394 

As a result, the number of training data should have decreased because we split it in the initial stage.

# Lab Work 4
Feature Extraction on Unstructured Data

In this work, we will attempt to extract features from an image. The process we will undertake is relatively simple. A comprehensive discussion regarding data preprocessing and feature extraction in image data will be covered in the digital image processing lectures.

### Step 0 - Install Library

In this work, we will use library called Pillow to perform some image processing. You can install Pillow directly on your notebook cell by this following command.

In [45]:
!pip install Pillow

### Step 1 - Load Image

In [51]:
from PIL import Image

img = Image.open('image.png')
img.show() # show the images

### Step 2 - Feature Extraction

In [50]:
# extact red, green, blue, channel
r, g, b = img.split()

# check the length of red feature
print(len(r.histogram()))

# Print the historgram of red channel
print(r.histogram())

256
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 8, 6, 10, 9, 20, 37, 32, 41, 53, 74, 103, 132, 179, 193, 242, 238, 336, 386, 404, 483, 607, 628, 771, 785, 1001, 1016, 1177, 1269, 1332, 1446, 1487, 1457, 1574, 1585, 1591, 1557, 1569, 1660, 1648, 1420, 1559, 1418, 1384, 1319, 1342, 1156, 1120, 955, 969, 828, 782, 752, 737, 719, 700, 628, 673, 587, 617, 610, 592, 557, 593, 552, 566, 582, 559, 571, 520, 664, 650, 618, 730, 594, 667, 675, 685, 771, 715, 667, 740, 744, 766, 765, 772, 817, 817, 744, 806, 760, 777, 812, 797, 799, 861, 814, 910, 907, 918, 888, 1011, 879, 996, 912, 952, 884, 1074, 977, 1073, 1040, 1216, 1250, 1403, 1534, 1639, 1682, 1776, 1874, 1769, 1582, 1743, 1441, 1477, 1483, 1409, 1437, 1449, 1389, 1479, 1592, 1655, 1657, 1666, 1857, 1896, 1813, 1979, 1814, 1956, 1928, 2055, 2012, 2303, 2333, 2670, 2787, 3232, 3154, 3476, 3424, 3516, 3102, 31